### PR TITLE
Use scheduledthreadpool instead of Timer #14

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

File: src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    private final Consumer<JobExecutionCommand> runCallback;
+
+    private final Predicate<JobDescriptor> runCheckFunction;
+
+    JobExecutionCommand( final JobDescriptor descriptor, final Consumer<JobExecutionCommand> runCallback )
+    {
+        this.descriptor = descriptor;
+        this.runCallback = runCallback;
+        this.runCount = new AtomicInteger( 0 );
+        this.runCheckFunction = ( JobDescriptor o ) -> o.getTimes().isEmpty() || this.runCount.get() < o.getTimes().get();
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java:33]: `o.getTimes().isEmpty()` could be evaluated earlier to construct `runCheckFunction = (...) -> true`

File: src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        try
+        {
+            if ( !runCheckFunction.test( this.descriptor ) )
+            {
+                this.runCallback.accept( this );
+                return;
+            }
+
+            this.doRun();
+            this.runCount.incrementAndGet();
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java:48]: `getAndIncrement` could be used instead straight in `runCheckFunction`

File: src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            this.runCallback.accept( this );
+        }
+        catch ( Exception e )
+        {
+            LOG.warn( String.format( "Error while running job [%s]", this.descriptor.getName() ), e );
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java:53]: You could save some CPU by using `{}` placeholders when logging

File: src/main/java/com/enonic/lib/cron/scheduler/JobScheduler.java [hunk 4]
```diff
@@ -1,19 +1,124 @@
...
+
+    private RecurringJobScheduler scheduler;
+
+    private Map<JobDescriptor, RecurringJob> tasks;
+
+    @Activate
+    public void setup( final BundleContext bundleContext )
+    {
+        this.scheduler = new RecurringJobScheduler( ApplicationKey.from( bundleContext.getBundle() ) + "-job-thread" );
+        this.tasks = Maps.newHashMap();
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/cron/scheduler/JobScheduler.java:33]: `ConcurrentHashMap` sounds cooler

File: src/main/java/com/enonic/lib/cron/scheduler/JobScheduler.java [hunk 5]
```diff
@@ -1,19 +1,124 @@
...
+            orElse( null );
+    }
+
+    public JobDescriptors list( final String jobNamePattern )
+    {
+        final JobDescriptors jobDescriptors = new JobDescriptors();
+
+        this.tasks.keySet().
+            stream().
+            filter( job -> StringUtils.isBlank( jobNamePattern ) || job.getName().matches( jobNamePattern ) ).
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/cron/scheduler/JobScheduler.java:87]: Don't use commons lib

File: src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java (hunk 2 above)
Comment (vbradnitski) [reply src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java:48]: there is also usage of `runCheckFunction` in `isFinished`

File: src/main/java/com/enonic/lib/cron/scheduler/JobScheduler.java (hunk 4 above)
Comment (vbradnitski) [reply src/main/java/com/enonic/lib/cron/scheduler/JobScheduler.java:33]: sure, will be done in issue-15

File: src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java (hunk 2 above)
Comment (rymsha) [reply src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java:50]: lets fix in #15

File: src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            }
+            else
+            {
+                this.doRun();
+                this.rerunCallback.accept( this );
+            }
+        }
+        catch ( Exception e )
+        {
+            LOG.warn( "Error while running job [{}]", this.descriptor.getName(), e );
```
Comment (anatol-sialitski) [inline src/main/java/com/enonic/lib/cron/scheduler/JobExecutionCommand.java:60]: `Log.error` instead of `Log.warn`

**Pull request merged**